hedera-bls-impl/src/main/java/com/hedera/platform/bls/impl/BLS12381BilinearMap.java

```diff
@@ -0,0 +1,118 @@
+    private static final Group SIGNATURE_GROUP = BLS12381Group1.getInstance();
+
+    /** The group of the bilinear map where BLS public keys reside */
+    private static final Group KEY_GROUP = BLS12381Group2.getInstance();
+
+    /** {@inheritDoc} */
+    @Override
+    public Field field() {
+        return FIELD;
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * <p>Since elements are smaller and faster to operate on, we are using {@link BLS12381Group1}
+     * as our signature group. More operations are performed with signatures than with keys
+     */
+    @Override
+    public Group signatureGroup() {
+        return SIGNATURE_GROUP;
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * <p>Since elements are larger and slower to operate on, we are using {@link BLS12381Group2} as
+     * our key group. Fewer operations are performed with keys than with signatures
+     */
+    @Override
+    public Group keyGroup() {
+        return KEY_GROUP;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public boolean comparePairing(
+            final GroupElement signatureElement1,
+            final GroupElement keyElement1,
+            final GroupElement signatureElement2,
+            final GroupElement keyElement2) {
+
+        if (!(signatureElement1 instanceof final BLS12381Group1Element signature1)) {
+            throw new IllegalArgumentException("signatureElement1 must be a BLS12381Group1Element");
+        }
+
+        if (!(keyElement1 instanceof final BLS12381Group2Element key1)) {
+            throw new IllegalArgumentException("keyElement1 must be a BLS12381Group2Element");
+        }
+
+        if (!(signatureElement2 instanceof final BLS12381Group1Element signature2)) {
+            throw new IllegalArgumentException("signatureElement2 must be a BLS12381Group1Element");
+        }
+
+        if (!(keyElement2 instanceof final BLS12381Group2Element key2)) {
+            throw new IllegalArgumentException("keyElement2 must be a BLS12381Group2Element");
+        }
+
+        return BLS12381Bindings.comparePairing(signature1, key1, signature2, key2);
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public byte[] displayPairing(
+            final GroupElement signatureElement, final GroupElement keyElement) {
+        if (!(signatureElement instanceof final BLS12381Group1Element signature)) {
+            throw new IllegalArgumentException("signatureElement must be a BLS12381Group1Element");
+        }
+
+        if (!(keyElement instanceof final BLS12381Group2Element key)) {
+            throw new IllegalArgumentException("keyElement must be a BLS12381Group2Element");
+        }
+
+        // display output is always this size
+        final byte[] output = new byte[1249];
+
+        final int errorCode;
+        if ((errorCode = BLS12381Bindings.pairingDisplay(signature, key, output)) != SUCCESS) {
+            throw new BLS12381Exception("pairingDisplay", errorCode);
+        }
```

Comment:
I see this pattern all over the place. Is there a reason why we couldn't instead do this? Doing the assignment in the if() statement doesn't save us any lines of code, but in my opinion makes the code harder to read.

```suggestion
        final int errorCode = BLS12381Bindings.pairingDisplay(signature, key, output);
        if (errorCode != SUCCESS) {
            throw new BLS12381Exception("pairingDisplay", errorCode);
        }
```

Reply:
It's a smell from a previous layout. Fixed throughout